src/gui/inputdevices/MouseInputHandler.cpp

```diff
@@ -111,21 +113,22 @@ auto MouseInputHandler::changeTool(InputEvent const& event) -> bool {
     Settings* settings = this->inputContext->getSettings();
     ToolHandler* toolHandler = this->inputContext->getToolHandler();
     GtkXournal* xournal = this->inputContext->getXournal();
+    bool toolChanged = false;
+    bool pressed = false;
+    bool configChanged = true;
 
-    ButtonConfig* cfg = nullptr;
-    if (modifier2 /* Middle Button */ && !xournal->selection) {
-        cfg = settings->getMiddleButtonConfig();
-    } else if (modifier3 /* Right Button */ && !xournal->selection) {
-        cfg = settings->getRightButtonConfig();
-    }
+    if (modifier2 /* Middle Button */ && !xournal->selection)
+        InputUtils::applyButton(toolHandler, settings, Buttons::BUTTON_MIDDLE, pressed, toolChanged, configChanged);
+    else if (modifier3 /* Right Button */ && !xournal->selection)
+        InputUtils::applyButton(toolHandler, settings, Buttons::BUTTON_RIGHT, pressed, toolChanged, configChanged);
 
-    if (cfg && cfg->getAction() != TOOL_NONE) {
-        toolHandler->pointCurrentToolToButtonTool();
-        cfg->acceptActions(toolHandler);
-    } else {
-        toolHandler->pointCurrentToolToToolbarTool();
+    if (!pressed || !configChanged) {
+        toolChanged |= toolHandler->pointCurrentToolToToolbarTool();
```

Comment:
This is a direct result of applying the button. Hence it should be part of this method. Or why did you leave this here?

Reply:
I left it at this place because the logic is actually a little bit more subtle. But I take that comment that the structure is not clear enough yet :sweat_smile: 
I think I found a quite clear solution now.

Reply:
> I left it at this place because the logic is actually a little bit more subtle.

Please add comments until this becomes clear. Left this way I would simplify this when I next touch the code. Another good reason to add comments 😉

Reply:
Did you check the latest version I pushed to this PR?
I think it's quite clean there, or would you still add comments to it?

```cpp
void MouseInputHandler::changeTool(InputEvent const& event) {
    Settings* settings = this->inputContext->getSettings();
    ToolHandler* toolHandler = this->inputContext->getToolHandler();
    GtkXournal* xournal = this->inputContext->getXournal();
    bool isClickOnSelection = this->inputContext->getXournal()->selection;
    bool toolChanged = false;

    if (modifier2 && !isClickOnSelection)
        toolChanged = InputUtils::applyButton(toolHandler, settings, Button::BUTTON_MOUSE_MIDDLE);
    else if (modifier3 && !isClickOnSelection)
        toolChanged = InputUtils::applyButton(toolHandler, settings, Button::BUTTON_MOUSE_RIGHT);
    else
        toolChanged = toolHandler->pointActiveToolToToolbarTool();

    if (toolChanged)
        toolHandler->fireToolChanged();
}
```

Reply:
Fixed in 38d509654dd5b017f84bf8b17f59b0a907d83638. Please resolve if you like the solution :)